src/back/TicketManager/src/Core/TicketManager.Core.Services/Endpoints/Payments/FinishPaymentEndpoint.cs

```diff
@@ -0,0 +1,57 @@
+using FastEndpoints;
+using MassTransit;
+using Microsoft.EntityFrameworkCore;
+using TicketManager.Core.Contracts.Payments;
+using TicketManager.Core.Domain.Accounts;
+using TicketManager.Core.Domain.Events;
+using TicketManager.Core.Services.DataAccess;
+using TicketManager.Core.Services.Processes.Events;
+using TicketManager.Core.Services.Services.HttpClients;
+
+namespace TicketManager.Core.Services.Endpoints.Payments;
+
+public class FinishPaymentEndpoint : Endpoint<FinishPaymentRequest, FinishPaymentResponse>
+{
+    private readonly PaymentClient paymentClient;
+    private readonly CoreDbContext dbContext;
+    private readonly IBus bus;
+
+    public FinishPaymentEndpoint(PaymentClient paymentClient, CoreDbContext dbContext, IBus bus)
+    {
+        this.paymentClient = paymentClient;
+        this.dbContext = dbContext;
+        this.bus = bus;
+    }
+
+    public override void Configure()
+    {
+        Post("/payment/finish");
+        Roles(AccountRoles.User);
+    }
+
+    public override async Task HandleAsync(FinishPaymentRequest req, CancellationToken ct)
+    {
+        var status = await paymentClient.GetPaymentStatusAsync(new() { Id = req.PaymentId, }, ct);
+
+        if (status?.Status != PaymentStatusDto.Confirmed)
+        {
+            await bus.Publish(new UnlockSeatsForInvalidPayment { PaymentId = req.PaymentId }, ct);
+            await SendOkAsync(new FinishPaymentResponse(), ct);
+            return;
+        }
+
+        var doesPaymentReservationExist = await dbContext
+            .Sectors
+            .AnyAsync(s => s.SeatReservations.Any(sr => sr.PaymentId == req.PaymentId && sr.CreationDate + SeatReservation.ReservationLifetime >= DateTime.UtcNow), ct);
```

Comment:
That why reservation lives 5 minutes longer than payment. Of course it doesn't cover 100% cases, but 99.(9)% cases. But let's say this accuracy is enough. If someone pays for the ticket and doesn't get it, we can give him money back manually. It's not a big deal if this happen once per 4 years or even less often.

Also it has to be done that way, because if reservation is expired someone else is able to reserve those seats, so they don't belong anymore to person who reserved them before.